File: Graph/python/graph_utils.py
from __future__ import annotations

from typing import Dict, List, Tuple

Graph = Dict[str, List[str]]
Grid = List[List[str]]


def build_undirected_graph(edges: List[Tuple[str, str]]) -> Graph:
    graph: Graph = {}
    for a, b in edges:
        graph.setdefault(a, []).append(b)
        graph.setdefault(b, []).append(a)
    return graph


def in_bounds(grid: Grid, r: int, c: int) -> bool:
    return 0 <= r < len(grid) and 0 <= c < len(grid[0])


def pos_key(r: int, c: int) -> str:
    return f"{r},{c}"

File: Graph/php/GraphUtils.php
<?php

declare(strict_types=1);

function buildUndirectedGraph(array $edges): array
{
    $graph = [];
    foreach ($edges as [$a, $b]) {
        $graph[$a] ??= [];
        $graph[$b] ??= [];
        $graph[$a][] = $b;
        $graph[$b][] = $a;
    }
    return $graph;
}

function gridInBounds(array $grid, int $row, int $col): bool
{
    return $row >= 0 && $row < count($grid) && $col >= 0 && $col < count($grid[0]);
}

function gridPosKey(int $row, int $col): string
{
    return $row . ',' . $col;
}
